lib/extract/dl_esm_inf/netcdf/kernel_data_netcdf.f90

```diff
@@ -0,0 +1,448 @@
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine writes the value of a dl_esm_field (r2d_field)
+    !! to the NetCDF file. It takes the variable id from the corresponding
+    !! declaration.
+    !! @param[inout] this The instance of the PSyDataType.
+    !! @param[in] name The name of the variable (string).
+    !! @param[in] value The value of the variable.
+    subroutine WriteFieldDouble(this, name, value)
+        use netcdf
+        use field_mod, only : r2d_field
+        implicit none
+        class(PSyDataType), intent(inout) :: this
+        character(*), intent(in) :: name
+        type(r2d_field), intent(in) :: value
+        integer :: retval
+        retval = CheckError(nf90_put_var(this%ncid, this%var_id(this%next_var_index),  &
+                                         value%data(:,:)))
+        this%next_var_index = this%next_var_index + 1
+    end subroutine WriteFieldDouble
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine reads the values of a dl_esm_inf field (r2d_field).
+    !! It allocates a 2d-double precision array to store the read values
+    !! which is then returned to the caller. If the memory for the array can
+    !! not be allocated, the application will be stopped.
+    !! @param[inout] this The instance of the PSyDataType.
+    !! @param[in] name The name of the variable (string).
+    !! @param[out] value An allocatable, unallocated 2d-double precision array
+    !!             which is allocated here and stores the values read.
+    subroutine ReadFieldDouble(this, name, value)
+        use netcdf
+        implicit none
+
+        class(PSyDataType), intent(inout) :: this
+        character(*), intent(in) :: name
+        double precision, dimension(:,:), allocatable :: value
+        integer :: retval, varid
+        integer :: dim1_id, dim1
+        integer :: dim2_id, dim2, ierr
+        character(100) :: x
+
+        ! First query the dimensions of the original r2d_field from the
+        ! netcdf file
+        retval = CheckError(nf90_inq_dimid(this%ncid, trim(name//"dim1"), dim1_id))
+        retval = CheckError(nf90_inquire_dimension(this%ncid, dim1_id,  &
+                                                   len=dim1))
+        retval = CheckError(nf90_inq_dimid(this%ncid, name//"dim2", dim2_id))
+        retval = CheckError(nf90_inquire_dimension(this%ncid, dim2_id,  &
+                                                   len=dim2))
+        ! Allocate enough space to store the values to be read:
+        allocate(value(dim1, dim2), Stat=ierr)
+        if (ierr /= 0) then
+            print *,"Can not allocate array for ",name," of size ", &
+                  dim1,"x",dim2,"."
+            stop
+        endif
+        ! Initialise it with 0, so that an array comparison will work
+        ! even though e.g. boundary areas or so might not be set at all.
+        value = 0
```

Comment:
Slightly picky (but some compilers may produce a warning), this should be:

    value = 0.0d0

Reply:
Done.